File: backend/app/domains/admin/schemas/incidents.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

IncidentStatus = Literal["investigating", "identified", "monitoring", "resolved"]
IncidentSeverity = Literal["critical", "high", "medium", "low"]

_ACTIVE_STATUSES = frozenset({"investigating", "identified", "monitoring"})


class IncidentNoteEntry(BaseModel):
    id: UUID
    note: str
    status_change: str | None
    created_by_id: UUID | None
    created_at: datetime


class IncidentSummary(BaseModel):
    id: UUID
    organization_id: UUID
    title: str
    status: str
    severity: str
    affected_services: list[str]
    message: str | None
    is_public: bool
    started_at: datetime
    resolved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class IncidentDetail(IncidentSummary):
    notes: list[IncidentNoteEntry] = Field(default_factory=list)


class IncidentsListResponse(BaseModel):
    items: list[IncidentSummary]
    total: int
    page: int
    page_size: int


class CreateIncidentRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    severity: IncidentSeverity = "medium"
    affected_services: list[str] = Field(default_factory=list, max_length=20)
    message: str | None = Field(None, max_length=5000)
    is_public: bool = False
    started_at: datetime | None = None


class UpdateIncidentRequest(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=255)
    status: IncidentStatus | None = None
    severity: IncidentSeverity | None = None
    affected_services: list[str] | None = None
    message: str | None = None
    is_public: bool | None = None
    resolved_at: datetime | None = None


class AddIncidentNoteRequest(BaseModel):
    note: str = Field(..., min_length=1, max_length=5000)
    status_change: IncidentStatus | None = None


class ServiceStatusBanner(BaseModel):
    has_active_incident: bool
    has_active_maintenance: bool
    active_incident_count: int
    banner_message: str | None
    highest_severity: str | None


class ServiceStatusSnapshot(BaseModel):
    organization_id: UUID
    generated_at: datetime
    active_incidents: list[IncidentSummary]
    recently_resolved: list[IncidentSummary]
    open_failed_job_count: int
    banner: ServiceStatusBanner

File: frontend/src/lib/api/incidents.ts
import { apiRequest } from "@/lib/api/request";

export type IncidentStatus =
  | "investigating"
  | "identified"
  | "monitoring"
  | "resolved";
export type IncidentSeverity = "critical" | "high" | "medium" | "low";

export type IncidentNoteEntry = {
  id: string;
  note: string;
  status_change: string | null;
  created_by_id: string | null;
  created_at: string;
};

export type IncidentSummary = {
  id: string;
  organization_id: string;
  title: string;
  status: IncidentStatus;
  severity: IncidentSeverity;
  affected_services: string[];
  message: string | null;
  is_public: boolean;
  started_at: string;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
};

export type IncidentDetail = IncidentSummary & {
  notes: IncidentNoteEntry[];
};

export type IncidentsListResponse = {
  items: IncidentSummary[];
  total: number;
  page: number;
  page_size: number;
};

export type ServiceStatusBanner = {
  has_active_incident: boolean;
  has_active_maintenance: boolean;
  active_incident_count: number;
  banner_message: string | null;
  highest_severity: string | null;
};

export type ServiceStatusSnapshot = {
  organization_id: string;
  generated_at: string;
  active_incidents: IncidentSummary[];
  recently_resolved: IncidentSummary[];
  open_failed_job_count: number;
  banner: ServiceStatusBanner;
};

export type IncidentsQuery = {
  status?: IncidentStatus;
  severity?: IncidentSeverity;
  active_only?: boolean;
  page?: number;
  page_size?: number;
};

export type CreateIncidentRequest = {
  title: string;
  severity?: IncidentSeverity;
  affected_services?: string[];
  message?: string | null;
  is_public?: boolean;
  started_at?: string | null;
};

export type UpdateIncidentRequest = {
  title?: string;
  status?: IncidentStatus;
  severity?: IncidentSeverity;
  affected_services?: string[];
  message?: string | null;
  is_public?: boolean;
  resolved_at?: string | null;
};

export type AddIncidentNoteRequest = {
  note: string;
  status_change?: IncidentStatus | null;
};

export async function getStatusSnapshot(): Promise<ServiceStatusSnapshot> {
  return apiRequest<ServiceStatusSnapshot>("/admin/status");
}

export async function getStatusBanner(): Promise<ServiceStatusBanner> {
  return apiRequest<ServiceStatusBanner>("/status/banner");
}

export async function listIncidents(
  query?: IncidentsQuery,
): Promise<IncidentsListResponse> {
  const params = new URLSearchParams();
  if (query?.status) params.set("status", query.status);
  if (query?.severity) params.set("severity", query.severity);
  if (query?.active_only) params.set("active_only", "true");
  if (query?.page != null) params.set("page", String(query.page));
  if (query?.page_size != null) params.set("page_size", String(query.page_size));
  const qs = params.toString();
  return apiRequest<IncidentsListResponse>(
    `/admin/incidents${qs ? `?${qs}` : ""}`,
  );
}

export async function getIncident(incidentId: string): Promise<IncidentDetail> {
  return apiRequest<IncidentDetail>(`/admin/incidents/${incidentId}`);
}

export async function createIncident(
  body: CreateIncidentRequest,
): Promise<IncidentDetail> {
  return apiRequest<IncidentDetail>("/admin/incidents", {
    method: "POST",
    json: body,
  });
}

export async function updateIncident(
  incidentId: string,
  body: UpdateIncidentRequest,
): Promise<IncidentDetail> {
  return apiRequest<IncidentDetail>(`/admin/incidents/${incidentId}`, {
    method: "PATCH",
    json: body,
  });
}

export async function addIncidentNote(
  incidentId: string,
  body: AddIncidentNoteRequest,
): Promise<IncidentDetail> {
  return apiRequest<IncidentDetail>(`/admin/incidents/${incidentId}/notes`, {
    method: "POST",
    json: body,
  });
}
